hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala

```diff
@@ -51,16 +125,84 @@ class ServiceBackend() extends Backend {
 
   def defaultParallelism: Int = 10
 
-  def broadcast[T: ClassTag](_value: T): BroadcastValue[T] = new BroadcastValue[T] {
+  def broadcast[T: ClassTag](_value: T): BroadcastValue[T] = new BroadcastValue[T] with Serializable {
     def value: T = _value
   }
 
-  def parallelizeAndComputeWithIndex[T: ClassTag, U : ClassTag](collection: Array[T])(f: (T, Int) => U): Array[U] = {
+  def parallelizeAndComputeWithIndex(_backendContext: BackendContext, collection: Array[Array[Byte]])(f: (Array[Byte], Int) => Array[Byte]): Array[Array[Byte]] = {
+    val backendContext = _backendContext.asInstanceOf[ServiceBackendContext]
+
     val n = collection.length
-    val r = new Array[U](n)
+
+    val token = tokenUrlSafe(32)
+
+    log.info(s"parallelizeAndComputeWithIndex: nPartitions $n token $token")
+
+    val root = s"gs://$queryBucket/stages/$token"
+
+    log.info(s"parallelizeAndComputeWithIndex: token $token: writing f")
+
+    using(new ObjectOutputStream(queryFS.create(s"$root/f"))) { os =>
+      os.writeObject(f)
+    }
+
+    log.info(s"parallelizeAndComputeWithIndex: token $token: writing context offsets")
+
+    using(queryFS.createNoCompression(s"$root/context.offsets")) { os =>
+      var o = 0L
+      var i = 0
+      while (i < n) {
+        val len = collection(i).length
+        os.writeLong(o)
+        os.writeInt(len)
+        i += 1
+        o += len
+      }
+    }
+
+    log.info(s"parallelizeAndComputeWithIndex: token $token: writing contexts")
+
+    using(queryFS.createNoCompression(s"$root/contexts")) { os =>
+      collection.foreach { context =>
+        os.write(context)
+      }
+    }
+
+    val jobs = new Array[JObject](n)
     var i = 0
     while (i < n) {
-      r(i) = f(collection(i), i)
+      jobs(i) = JObject(
+          "always_run" -> JBool(false),
+          "image" -> JString(workerImage),
+          "mount_docker_socket" -> JBool(false),
+          "command" -> JArray(List(
+            JString("/bin/bash"),
+            JString("-c"),
+            JString(s"java -cp $$SPARK_HOME/jars/*:/hail.jar is.hail.backend.service.Worker $root $i"))),
+          "job_id" -> JInt(i),
+          "parent_ids" -> JArray(List()))
+      i += 1
+    }
+
+    log.info(s"parallelizeAndComputeWithIndex: token $token: running job")
+
+    val batchClient = BatchClient.fromSessionID(backendContext.sessionID)
+    // FIXME check return
+    val batch = batchClient.run(
+      JObject(
+        "billing_project" -> JString(backendContext.billingProject),
+        "n_jobs" -> JInt(n),
+        "token" -> JString(token)),
+      jobs)
```

Comment:
seems like we should assert that every job succeeded, right?

Reply:
I throw `RuntimeException` if `state != "success"`.  This is bare bones, there's going to be more work (1) getting a nice error message to the user (file not found, array index exception, etc.) and (2) logging well so we can debug system errors.